README.md
# node-red-flow2src
Now you can easily see git diff, use CLI editors, or third-party IDEs like Visual Studio Code to leverage Co-Pilot with your Node-RED based projects. This node will write template and function node source code properties to a src sub-folder adjacent to the project's flows file; and it will allow you to merge changes back into your flows when you're done editing.

## How to Use flow2src
Simply include the flow2src node on one of your project's flows. When enabling [Node-RED's project mode](https://nodered.org/docs/user-guide/projects/), the files can be easily managed within Node-RED itself. See the screenshot below for reference:

1) Include the flow2src node on one of your flows and open its property sheet.
2) Click the "Flow-to-Src" button; flow2src will generate the files. 
3) Click Node-RED's "Refresh changes" button in the history tab. 

![screenshot of node-maker](https://raw.github.com/steveorevo/node-red-flow2src/main/images/flow2src.jpg)

By default, flow2src will look for all flows and subflows for template and function nodes and create a src folder. Within the src folder are a number of subfolders of the same name as your flow tabs and subflows. Source code files with properly named
extensions (i.e. .js, .html, .json, .css, etc.) are written to the folders. The filename extensions are derived from the node type (i.e. function node creates .js files) or the 'Syntax Highlight' mode setting for [template nodes](https://nodered.org/docs/user-guide/nodes#template). You can override the filename extension by explicitly naming your node with an extension. Nodes that share the same name within a given flow will automatically be iterated, i.e. function_1.js, function_2.js, etc.

After making edits to the files in your src folder; use the "Src-to-Flow" button to merge changes back into your flow file. 

### Options
You can further customize which flows and subflows are exported and where they are exported using the following textbox options:

* Flows - specify a flow by name, or mulitple flows by seperating names by a comma. Use an asterick for all flows.
* Subflows - specify a subflow by name, or mulitple subflows by seperating names by a comma. Use an asterick for all subflows.
* Output Folder - By default the folder is named 'src'; you can specify a different path relative to your project's flow file.

Lastly, you can invoke the "Flow-to-Src" button automatically using the checkbox for "Automatiaclly Flow-to-Src on Deploys".

## How to Install
You can install flow2src via [Node-RED's built in palette manager](https://nodered.org/docs/user-guide/editor/palette/manager) and searching for `node-red-flow2src`

or 

Run the following command in your Node-RED user directory (typically ~/.node-red):

    npm install @steveorevo/node-red-flow2src

The flow2src node will appear in the palette under the common group.

## Building
The flow2src node was built using the same author's [Node-Maker](https://github.com/steveorevo/node-maker) project. Included in this repo is the flow2src.json file containing the flow needed to build flow2src. Simply import this flow into Node-RED and click the associated inject button. The current version of node-red-flow2src folder will appear in your Node-RED's node_modules folder. 

## Support the Creator
You can help Steveorevo's open source development endeavors by donating any amount. Your donation, no matter how large or small helps pay for his time and resources to create MIT and GPL licensed projects that you and the world can benefit from. Click the link below to donate today :)
<div>
         

[<kbd> <br> Donate to this Project <br> </kbd>][KBD]


</div>


<!---------------------------------------------------------------------------->

[KBD]: https://steveorevo.com/donate

https://steveorevo.com/donate


CHANGELOG.md
# Changelog

## [1.1.0] - 2022-11-29

Automatic flows reload added to Src-to-Flow

## [1.0.5] - 2022-11-27

_Initial release._



flow2src/flow2src.js
module.exports = function(RED) {
    function flow2src(config) {
        RED.nodes.createNode(this, config);
        var node = this;
        node.on('input', function (msg) {
            if (!msg.hasOwnProperty('srcFolder')) return;
            if (msg.srcFolder.trim() == '') msg.srcFolder = 'src';

            // String manipulation functions
            (function () {
                String.prototype.delRightMost = function(sFind) {
                    for (var i = this.length; i >= 0; i = i - 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(0, f);
                            break;
                        }
                    }
                    return this;
                };
                String.prototype.getRightMost = function(sFind) {
                    for (var i = this.length; i >= 0; i = i - 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(f + sFind.length, f + sFind.length + this.length);
                        }
                    }
                    return this;
                };
                String.prototype.delLeftMost = function(sFind) {
                    for (var i = 0; i < this.length; i = i + 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(f + sFind.length, f + sFind.length + this.length);
                            break;
                        }
                    }
                    return this;
                };
                String.prototype.getLeftMost = function(sFind) {
                    for (var i = 0; i < this.length; i = i + 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(0, f);
                            break;
                        }
                    }
                    return this;
                };
            })();

            // Get the flowFile
            const fs = require('fs');
            let flowFile = RED.settings.userDir;
            try{
                if (RED.settings.get('editorTheme').projects.enabled) {
                    let project = RED.settings.get('projects').activeProject;
                    let package_json = flowFile + '/projects/' + project + '/package.json';

                    // Read the package.json for the flowfile
                    let pk = JSON.parse(fs.readFileSync(package_json).toString());
                    flowFile += '/projects/' + project + '/' + pk['node-red']['settings']['flowFile'];
                } else {
                    flowFile += '/' + RED.settings.flowFile;
                }
            } catch(e) {
                node.error(e);
                return;
            }

            // Read and parse the flow file
            let ff = JSON.parse(fs.readFileSync(flowFile).toString());
            let path = flowFile.delRightMost('/') + '/' + msg.srcFolder;

            // Write the relevant flow properties to the src folder
            if (msg.action == 'flow2src') {
                try {

                    // Gather flows and subflows as an array
                    let incFlows = [];
                    if (config.incFlows.trim() != '*') {
                        incFlows = config.incFlows.split(',').map(function (f) { return f.trim() });
                    } else {
                        incFlows = null;
                    }
                    let incSubflows = [];
                    if (config.incSubflows.trim() != '*') {
                        incSubflows = config.incSubflows.split(',').map(function (f) { return f.trim() });
                    } else {
                        incSubflows = null;
                    }

                    // Gather flow ids, folder safe names, and nodes for analysis in a single pass
                    let theNodes = [];
                    let theFolders = [];
                    let theIDs = [];
                    ff.forEach(function (obj) {

                        // Check for matching flows
                        if (obj.type == 'tab') {
                            if (incFlows != null) {
                                if (incFlows.indexOf(obj.label) == -1) {
                                    return;
                                }
                            }
                            theFolders.push(obj.label.replace(/[^a-z0-9]/gi, '_'));
                            theIDs.push(obj.id);
                        }

                        // Check for matching subflows
                        if (obj.type == 'subflow') {
                            if (incSubflows != null) {
                                if (incSubflows.indexOf(obj.name) == -1) {
                                    return;
                                }
                            }
                            theFolders.push(obj.name.replace(/[^a-z0-9]/gi, '_'));
                            theIDs.push(obj.id);
                        }
                        // Gather nodes and templates
                        if (obj.type == 'template' || obj.type == 'function') {
                            theNodes.push(obj);
                        }
                    });

                    // Narrow the nodes to just the flows and subflows we're interested in
                    let existingFiles = [];
                    let srcNodes = [];
                    theNodes.forEach(function (obj) {
                        if (theIDs.indexOf(obj.z) == -1) return;
                        
                        // Determine the filename extension
                        let ext = '';
                        if (obj.type == 'template') {
                            ext = obj.format.toLowerCase();
                            if (ext == 'handlebars' || ext == 'text') {
                                ext = '';
                            }
                            if (ext == 'javascript') {
                                ext = 'js';
                            }
                            if (ext != '') {
                                ext = '.' + ext;
                            }
                        }
                        if (obj.type == 'function') {
                            ext = '.js';
                        }
                        let fname = obj.name.replace(/[^a-z0-9]/gi, '_');
                        if (fname == '') {
                            fname = 'untitled';
                        }

                        // Use existing extension in filename
                        if (fname.indexOf('.') != -1) {
                            ext = '.' + fname.getRightMost('.');
                            fname = fname.delRightMost('.');
                        }
                        let file = path + '/' + theFolders[theIDs.indexOf(obj.z)] + '/' + fname + ext;
                        let i = 2;

                        // Iterate existing filenames
                        while (existingFiles.indexOf(file) != -1) {
                            file = path + '/' + theFolders[theIDs.indexOf(obj.z)] + '/' + fname + i.toString() + ext;
                            i++;
                        }
                        obj.srcFiles = [];
                        if (obj.type == 'template') {
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'template',
                                file: file
                            });
                            existingFiles.push(file);
                        } else if (obj.type == 'function') {
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'func',
                                file: file
                            });
                            existingFiles.push(file);

                            // Record function On Start and On Stop too
                            let onStartFile = file;
                            ext = onStartFile.getRightMost('.');
                            onStartFile = onStartFile.delRightMost('.') + '_initialize' + ext;
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'initialize',
                                file: onStartFile
                            });
                            let onStopFile = file;
                            ext = onStopFile.getRightMost('.');
                            onStopFile = onStopFile.delRightMost('.') + '_finalize' + ext;
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'finalize',
                                file: onStopFile
                            });
                        }
                        srcNodes.push(obj);
                    });

                    // Remove prior src folder
                    fs.rmSync(path, { recursive: true, force: true });

                    // Write the nodes to the src folder and record the manifest
                    let manifest = [];
                    srcNodes.forEach(function (obj) {
                        obj.srcFiles.forEach(function (sF) {

                            // Create the src path
                            let sFPath = sF.file.delRightMost('/');
                            if (!fs.existsSync(sFPath)) {
                                fs.mkdirSync(sFPath, { recursive: true });
                            }

                            // Write the given file
                            if (obj[sF.property] != '') {
                                fs.writeFileSync(sF.file, obj[sF.property]);
                                sF.file = sF.file.delLeftMost(path + '/');
                                manifest.push(sF);
                            }
                        });
                    });

                    // Write the manifest to the src folder
                    fs.writeFileSync(path + '/manifest.json', JSON.stringify(manifest, null, 4));
                    node.status({ fill: "green", shape: "dot", text: "updated files" });
                    setTimeout(function() {
                        node.status({});
                    }, 5000);
                } catch(e) {
                    node.error(e);
                }
            }

            // Read src folder and update the flow
            if (msg.action == 'src2flow') {
                try {

                    // Load the manifest
                    if (!fs.existsSync(path + '/manifest.json')) return;
                    let mn = JSON.parse(fs.readFileSync(path + '/manifest.json').toString());

                    ff.forEach(function (obj) {
                        mn.forEach(function(item) {
                            if (item.id != obj.id) return;
                            
                            // Update the content from the external file
                            let file = fs.readFileSync(path + '/' + item.file).toString();
                            obj[item.property] = file;
                        });
                    });

                    // Update the flow file
                    fs.writeFileSync(flowFile, JSON.stringify(ff, null, 4));
                } catch(e) {
                    node.error(e);
                }
            }
        });

        // Automatic flow2src on deploys
        if (config.chkAutoFlow2Src) {
            node.receive({action:"flow2src", srcFolder: config.srcFolder});
        }
    }
    RED.httpAdmin.post("/flow2src/:id", RED.auth.needsPermission("inject.write"), function (req, res) {
        var node = RED.nodes.getNode(req.params.id);
        if (node != null) {
            try {
                if (req.body) {
                    node.receive(req.body);
                } else {
                    node.receive();
                }
                res.sendStatus(200);
            } catch (err) {
                res.sendStatus(500);
                node.error(RED._("flow2src.failed", { error: err.toString() }));
            }
        } else {
            res.sendStatus(404);
        }
    });
    RED.nodes.registerType('flow2src', flow2src);
}

src/start_node/function_1.js
msg.node_name = env.get('node_name');
msg.node_version = env.get('node_version');
msg.node_description = env.get('node_description');
msg.node_folder = env.get('HOME') + '/.node-red/node_modules/node-red-' + msg.node_name + '/';
let k = env.get('keywords');
k = '"' + k.replaceAll(' ', '').replaceAll(',', '","') + '"';
msg.node_category = env.get('node_category');
msg.node_icon = env.get('node_icon');
msg.node_color = env.get('node_color');
msg.keywords = k;
msg.author_name = env.get('author_name');
msg.author_email = env.get('author_email');
msg.github_handle = env.get('github_handle');
msg.package_json = msg.node_folder + 'package.json';
msg.gitignore = msg.node_folder + '.gitignore';
msg.license = msg.node_folder + 'LICENSE';
msg.readme = msg.node_folder + 'README.md';
msg.nodename_js = msg.node_folder + msg.node_name + '/' + msg.node_name + '.js';
msg.nodename_html = msg.node_folder + msg.node_name + '/' + msg.node_name + '.html';
msg.edit_dialog = [];
msg._oneditprepare = [];
msg._oneditresize = [];
msg._oneditsave = [];
msg._oneditcancel = [];
msg._othercode = {};
msg.defaults = [];
msg.row_width = env.get('row_width');
msg.numinputs = env.get('numinputs');
msg.numoutputs = env.get('numoutputs');
msg.inc_node_name = env.get('inc_node_name');
msg.nodemakerhtml = '';
msg.nodemakerjs = '';
msg.runtimecode = '';
if (msg.inc_node_name) {
    msg.inc_node_name = '    <div class="form-row">' +"\n";
    msg.inc_node_name += '        <label for="node-input-name"><i class="icon-tag"></i> Name</label>' + "\n";
    msg.inc_node_name += '        <input type="text" id="node-input-name" placeholder="Name">' + "\n";
    msg.inc_node_name += "    </div>";
}else{
    msg.inc_node_name = '';
}
return msg;


src/flow2src/nodemakerjs.js
module.exports = function(RED) {
    function {{node_name}}(config) {
        RED.nodes.createNode(this, config);
        var node = this;
        node.on('input', function (msg) {
            if (!msg.hasOwnProperty('srcFolder')) return;
            if (msg.srcFolder.trim() == '') msg.srcFolder = 'src';

            // String manipulation functions
            (function () {
                String.prototype.delRightMost = function(sFind) {
                    for (var i = this.length; i >= 0; i = i - 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(0, f);
                            break;
                        }
                    }
                    return this;
                };
                String.prototype.getRightMost = function(sFind) {
                    for (var i = this.length; i >= 0; i = i - 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(f + sFind.length, f + sFind.length + this.length);
                        }
                    }
                    return this;
                };
                String.prototype.delLeftMost = function(sFind) {
                    for (var i = 0; i < this.length; i = i + 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(f + sFind.length, f + sFind.length + this.length);
                            break;
                        }
                    }
                    return this;
                };
                String.prototype.getLeftMost = function(sFind) {
                    for (var i = 0; i < this.length; i = i + 1) {
                        var f = this.indexOf(sFind, i);
                        if (f != -1) {
                            return this.substring(0, f);
                            break;
                        }
                    }
                    return this;
                };
            })();

            // Get the flowFile
            const fs = require('fs');
            let flowFile = RED.settings.userDir;
            try{
                if (RED.settings.get('editorTheme').projects.enabled) {
                    let project = RED.settings.get('projects').activeProject;
                    let package_json = flowFile + '/projects/' + project + '/package.json';

                    // Read the package.json for the flowfile
                    let pk = JSON.parse(fs.readFileSync(package_json).toString());
                    flowFile += '/projects/' + project + '/' + pk['node-red']['settings']['flowFile'];
                } else {
                    flowFile += '/' + RED.settings.flowFile;
                }
            } catch(e) {
                node.error(e);
                return;
            }

            // Read and parse the flow file
            let ff = JSON.parse(fs.readFileSync(flowFile).toString());
            let path = flowFile.delRightMost('/') + '/' + msg.srcFolder;

            // Write the relevant flow properties to the src folder
            if (msg.action == 'flow2src') {
                try {

                    // Gather flows and subflows as an array
                    let incFlows = [];
                    if (config.incFlows.trim() != '*') {
                        incFlows = config.incFlows.split(',').map(function (f) { return f.trim() });
                    } else {
                        incFlows = null;
                    }
                    let incSubflows = [];
                    if (config.incSubflows.trim() != '*') {
                        incSubflows = config.incSubflows.split(',').map(function (f) { return f.trim() });
                    } else {
                        incSubflows = null;
                    }

                    // Gather flow ids, folder safe names, and nodes for analysis in a single pass
                    let theNodes = [];
                    let theFolders = [];
                    let theIDs = [];
                    ff.forEach(function (obj) {

                        // Check for matching flows
                        if (obj.type == 'tab') {
                            if (incFlows != null) {
                                if (incFlows.indexOf(obj.label) == -1) {
                                    return;
                                }
                            }
                            theFolders.push(obj.label.replace(/[^a-z0-9]/gi, '_'));
                            theIDs.push(obj.id);
                        }

                        // Check for matching subflows
                        if (obj.type == 'subflow') {
                            if (incSubflows != null) {
                                if (incSubflows.indexOf(obj.name) == -1) {
                                    return;
                                }
                            }
                            theFolders.push(obj.name.replace(/[^a-z0-9]/gi, '_'));
                            theIDs.push(obj.id);
                        }
                        // Gather nodes and templates
                        if (obj.type == 'template' || obj.type == 'function') {
                            theNodes.push(obj);
                        }
                    });

                    // Narrow the nodes to just the flows and subflows we're interested in
                    let existingFiles = [];
                    let srcNodes = [];
                    theNodes.forEach(function (obj) {
                        if (theIDs.indexOf(obj.z) == -1) return;
                        
                        // Determine the filename extension
                        let ext = '';
                        if (obj.type == 'template') {
                            ext = obj.format.toLowerCase();
                            if (ext == 'handlebars' || ext == 'text') {
                                ext = '';
                            }
                            if (ext == 'javascript') {
                                ext = 'js';
                            }
                            if (ext != '') {
                                ext = '.' + ext;
                            }
                        }
                        if (obj.type == 'function') {
                            ext = '.js';
                        }
                        let fname = obj.name.replace(/[^a-z0-9]/gi, '_');
                        if (fname == '') {
                            fname = 'untitled';
                        }

                        // Use existing extension in filename
                        if (fname.indexOf('.') != -1) {
                            ext = '.' + fname.getRightMost('.');
                            fname = fname.delRightMost('.');
                        }
                        let file = path + '/' + theFolders[theIDs.indexOf(obj.z)] + '/' + fname + ext;
                        let i = 2;

                        // Iterate existing filenames
                        while (existingFiles.indexOf(file) != -1) {
                            file = path + '/' + theFolders[theIDs.indexOf(obj.z)] + '/' + fname + i.toString() + ext;
                            i++;
                        }
                        obj.srcFiles = [];
                        if (obj.type == 'template') {
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'template',
                                file: file
                            });
                            existingFiles.push(file);
                        } else if (obj.type == 'function') {
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'func',
                                file: file
                            });
                            existingFiles.push(file);

                            // Record function On Start and On Stop too
                            let onStartFile = file;
                            ext = onStartFile.getRightMost('.');
                            onStartFile = onStartFile.delRightMost('.') + '_initialize' + ext;
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'initialize',
                                file: onStartFile
                            });
                            let onStopFile = file;
                            ext = onStopFile.getRightMost('.');
                            onStopFile = onStopFile.delRightMost('.') + '_finalize' + ext;
                            obj.srcFiles.push({
                                id: obj.id,
                                property: 'finalize',
                                file: onStopFile
                            });
                        }
                        srcNodes.push(obj);
                    });

                    // Remove prior src folder
                    fs.rmSync(path, { recursive: true, force: true });

                    // Write the nodes to the src folder and record the manifest
                    let manifest = [];
                    srcNodes.forEach(function (obj) {
                        obj.srcFiles.forEach(function (sF) {

                            // Create the src path
                            let sFPath = sF.file.delRightMost('/');
                            if (!fs.existsSync(sFPath)) {
                                fs.mkdirSync(sFPath, { recursive: true });
                            }

                            // Write the given file
                            if (obj[sF.property] != '') {
                                fs.writeFileSync(sF.file, obj[sF.property]);
                                sF.file = sF.file.delLeftMost(path + '/');
                                manifest.push(sF);
                            }
                        });
                    });

                    // Write the manifest to the src folder
                    fs.writeFileSync(path + '/manifest.json', JSON.stringify(manifest, null, 4));
                    node.status({ fill: "green", shape: "dot", text: "updated files" });
                    setTimeout(function() {
                        node.status({});
                    }, 5000);
                } catch(e) {
                    node.error(e);
                }
            }

            // Read src folder and update the flow
            if (msg.action == 'src2flow') {
                try {

                    // Load the manifest
                    if (!fs.existsSync(path + '/manifest.json')) return;
                    let mn = JSON.parse(fs.readFileSync(path + '/manifest.json').toString());

                    ff.forEach(function (obj) {
                        mn.forEach(function(item) {
                            if (item.id != obj.id) return;
                            
                            // Update the content from the external file
                            let file = fs.readFileSync(path + '/' + item.file).toString();
                            obj[item.property] = file;
                        });
                    });

                    // Update the flow file
                    fs.writeFileSync(flowFile, JSON.stringify(ff, null, 4));
                } catch(e) {
                    node.error(e);
                }
            }
        });

        // Automatic flow2src on deploys
        if (config.chkAutoFlow2Src) {
            node.receive({action:"flow2src", srcFolder: config.srcFolder});
        }
    }
    RED.httpAdmin.post("/flow2src/:id", RED.auth.needsPermission("inject.write"), function (req, res) {
        var node = RED.nodes.getNode(req.params.id);
        if (node != null) {
            try {
                if (req.body) {
                    node.receive(req.body);
                } else {
                    node.receive();
                }
                res.sendStatus(200);
            } catch (err) {
                res.sendStatus(500);
                node.error(RED._("flow2src.failed", { error: err.toString() }));
            }
        } else {
            res.sendStatus(404);
        }
    });
    RED.nodes.registerType('{{node_name}}', {{node_name}});
}